src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Utilities/MetadataExtensions.cs

```diff
@@ -0,0 +1,80 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+
+using Microsoft.VisualStudio.ProjectSystem.Properties;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Utilities
+{
+    internal static class MetadataExtensions
+    {
+        /// <summary>
+        /// Finds the resolved reference item for a given unresolved reference.
+        /// </summary>
+        /// <param name="projectRuleSnapshot">Resolved reference project items snapshot to search.</param>
+        /// <param name="itemSpec">The unresolved reference item name.</param>
+        /// <returns>The key is item name and the value is the metadata dictionary.</returns>
+        public static IImmutableDictionary<string, string> GetProjectItemProperties(this IProjectRuleSnapshot projectRuleSnapshot, string itemSpec)
+        {
+            Requires.NotNull(projectRuleSnapshot, nameof(projectRuleSnapshot));
+            Requires.NotNullOrEmpty(itemSpec, nameof(itemSpec));
+
+            projectRuleSnapshot.Items.TryGetValue(itemSpec, out IImmutableDictionary<string, string> properties);
+
+            return properties;
+        }
+
+        public static bool TryGetStringProperty(this IImmutableDictionary<string, string> properties, string key, out string stringValue)
+        {
+            if (properties != null &&
+                properties.TryGetValue(key, out stringValue) &&
+                !string.IsNullOrEmpty(stringValue))
```

Comment:
Why would we have a null or empty string in there to begin with?

Reply:
I'm not sure why, but this check is pretty common amongst existing code that reads properties from dictionaries. Would `<EmptyElement></EmptyElement>` not produce an empty property?

Reply:
That still doesn't explain the null case for me

---
In reply to: [231937916](https://github.com/dotnet/project-system/pull/4242#discussion_r231937916) [](ancestors = 231937916)

Reply:
And if the property is an empty string why is it incorrect to return it?

---
In reply to: [231938983](https://github.com/dotnet/project-system/pull/4242#discussion_r231938983) [](ancestors = 231938983,231937916)

Reply:
What would be better here?

Reply:
I don't know, I am moistly asking about the intended semantics of this method.  Its clear that if we have a value for a key that is null or an empty string we consider that the wrong result and return false, but it is unclear to me why 

---
In reply to: [231939349](https://github.com/dotnet/project-system/pull/4242#discussion_r231939349) [](ancestors = 231939349)

Reply:
The intent of the `Get*` variants is that they return `null` if no sensible value could be obtained, and the caller can then `??` a default in. It's the more convenient form. You can `?? string.Empty` if you want.

I'll document these methods.

Reply:
Yes but if someone added an empty string or even null intentially to this dictionary why are those values not sensible? Essentially I am asking why this is wrong:

```C#
if (properties != null && properties.TryGetValue(key, out stringValue) {
   return true;
}
```

Reply:
The decision was based on existing usages. Many test for empty strings and do something else if they're observed. This just makes that simpler. My understanding is that in MSBuild and empty string is equivalent to a null, as far as property values are concerned. Correct me if I'm mistaken.